src/profile/webid.ts

```diff
@@ -71,28 +71,38 @@ export async function getProfileAll<T extends SolidDataset & WithResourceInfo>(
   const profileDocument =
     webIdProfile ?? ((await getSolidDataset(webId, { fetch })) as T);
 
-  const profilesLinkedFrom = getThingAll(profileDocument)
-    .filter((thing) => getIriAll(thing, foaf.primaryTopic).length > 0)
-    // This assumes that getSourceIri returns the IRI where the profile document
-    // has actually been fetched, which may differ from the WebID.
-    .filter((thing) => asIri(thing) !== getSourceIri(profileDocument))
-    .map((primaryThing) => getSolidDataset(asIri(primaryThing), { fetch }));
+  const profilesLinkedFrom = await Promise.all(
+    getThingAll(profileDocument)
+      .filter((thing) => getIriAll(thing, foaf.primaryTopic).length > 0)
+      // This assumes that getSourceIri returns the IRI where the profile document
+      // has actually been fetched, which may differ from the WebID.
+      .filter((thing) => asIri(thing) !== getSourceIri(profileDocument))
+      .map((primaryThing) => getSolidDataset(asIri(primaryThing), { fetch }))
+  );
+
+  const fetchedProfilesIri = profilesLinkedFrom.map(getSourceIri);
 
   let profilesLinkedTo: Array<Promise<SolidDataset & WithResourceInfo>> = [];
   const webIdThing = getThing(profileDocument, webId);
   if (webIdThing !== null) {
     profilesLinkedTo = getIriAll(webIdThing, foaf.isPrimaryTopicOf)
       .filter(
```

Comment:
```suggestion
  const { fetch, webIdProfile = await getSolidDataset(webId, { fetch }) as T } = options;

  const webIdThing = getThing(profileDocument, webId);

  const altProfilesIri = getThingAll(profileDocument)
    .filter(thing => getIriAll(thing, foaf.primaryTopic).length > 0)
    .map(asIri)
    .concat(
      webIdThing ? getIriAll(webIdThing, foaf.isPrimaryTopicOf) : []        
    )
    .filter(profileIri => profileIri !== getSourceIri(profileDocument));
    
  return {
    webIdProfile,
    altProfileAll: await Promise.all(
      [...new Set(altProfilesIri)]
        .map(uniqueProfileIri => getSolidDataset(asIri(primaryThing), { fetch })
    )
  };
```

Reply:
Maybe a bit less repetition if starting by extracting the IRIs and destructured default assignment.
(I just wrote this here, so needs to be tested ^^,)

Reply:
I'd go similar to the above suggestion, but also de-duplicating between the webIdProfile and the alt profiles, essentially an `IrisToFetch` and then de-structure that into the shapes we want to return. I'd also lift the `await Promise.all` up into a constant.

Also, `Array.from(new Set(altProfilesIri))` instead of the spread.

There's also a question of how should this fail? Should it only fail if the webIdProfile fails to load, or should it fail if all the altProfiles fail to load?

Reply:
I'm not sure we need de-duplicating between the webid profile and the alt ones, as that would involve the triples in the shape `<webid (foaf:primaryTopic|foaf:isPrimaryTopicOf) webid>`, which doesn't really make sense.

I think it should fail if any profile, the WebID one or the alt ones, fail to load. That's the current behaviour.

Reply:
920ad00b

Reply:
> Also, Array.from(new Set(altProfilesIri)) instead of the spread.

@ThisIsMissEm I think the spread is the most idiomatic and concise just like destructured assignments of options. Obviously it's a matter of preference, but I think it's just more readable/scanable/visually easy to parse. That's always the angle I come from in my preference of syntax.

So a few that I love to use here represented:

```
// Destructured assignment of options with defaults
const options = { a: 0 };
const { a = 1, b = 1 } = options;
```

```
// Variable assignment in objects (not sure what the precise name is)
const a = 1;
const x = { a };
```

```
// Unique values in array
const a = [ 1, 1, 2, 3 ];
const unique = [ ...new Set(a) ];
```

Reply:
Aaah, I didn't realize the destructured assignment had a default value ^^. Fixed that in 581bfe3d.

Reply:
However, I think my preference goes to `Array.from()` rather than destructuring too from a readability standpoint, but it's purely a personal opinion. Are there additional elements to consider in this choice ?